R/MsBackendCached.R

```diff
@@ -0,0 +1,616 @@
+#'   the responsibility of the extending backend to add/provide these.
+#'
+#' - `[`: subsets the cached data. Parameter `i` needs to be an `integer`
+#'   vector.
+#'
+#' - `$`, `$<-`: access or set/add a single spectrum variable (column) in the
+#'   backend.
+#'
+#'
+#' @param columns For `spectraData`: `character` with the names of the spectra
+#'     variables to retrieve.
+#'
+#' @param data For `backendInitialize`: (optional) `data.frame` with cached
+#'     values. The number of rows (and their order) has to match the number of
+#'     spectra.
+#'
+#' @param drop For `[`: not considered.
+#'
+#' @param i For `[`: `integer` with the indices to subset the object.
+#'
+#' @param j For `[`: ignored.
+#'
+#' @param name For `$<-`: the name of the spectra variable to set.
+#'
+#' @param nspectra For `backendInitialize`: `integer` with the number of
+#'     spectra.
+#'
+#' @param object A `MsBackendCached` object.
+#'
+#' @param spectraVariables For `backendInitialize`: `character` with the names
+#'     of the spectra variables that are provided by the extending backend.
+#'     For `selectSpectraVariables`: `character` specifying the spectra
+#'     variables to keep.
+#'
+#' @param use.names For `lengths`: whether spectrum names should be used.
+#'
+#' @param value replacement value for `<-` methods. See individual
+#'     method description or expected data type.
+#'
+#' @param x A `MsBackendCached` object.
+#'
+#' @param ... ignored
+#'
+#' @author Johannes Rainer
+#'
+#' @name MsBackendCached
+#'
+#' @return See documentation of respective function.
+#'
+#' @seealso [MsBackend] for the documentation of MS backends.
+#'
+#' @md
+#'
+#' @exportClass MsBackendCached
+NULL
+
+setClass(
+    "MsBackendCached",
+    contains = "MsBackend",
+    slots = c(
+        spectraVariables = "character",
+        localData = "data.frame",
+        nspectra = "integer"),
+    prototype = prototype(
+        spectraVariables = character(),
+        localData = data.frame(),
+        nspectra = 0L,
+        readonly = FALSE, version = "0.1"))
+
+.valid_local_data <- function(x, y) {
+    if (nrow(x) && nrow(x) != y)
+        "Number of rows in local data and number of spectra don't match"
+    else NULL
+}
+
+#' @rdname MsBackendCached
+#'
+#' @export
+MsBackendCached <- function() {
+    new("MsBackendCached")
+}
+
+setValidity("MsBackendCached", function(object) {
+    msg <- c(.valid_local_data(object@localData, object@nspectra))
+    ## For local data that are core spectra variables, check for valid data type
+    msg <- c(msg, .valid_column_datatype(object@localData))
+    if (is.null(msg)) TRUE
+    else msg
+})
+
+#' @rdname MsBackendCached
+setMethod("backendInitialize", "MsBackendCached",
+          function(object, data = data.frame(), nspectra = 0L,
+                   spectraVariables = character(), ...) {
+              if (nrow(data) == 0L)
+                  data <- data.frame(matrix(ncol = 0, nrow = nspectra))
```

Comment:
or `data.frame(row.names = seq_len(nspectra))`

Reply:
That's nice. Thanks for the hint.